sdk/burdock/query/sql/private/query.py

```diff
@@ -52,17 +53,40 @@ def _apply_noise(self, subquery, query, syms, types, sens, srs):
             name, sym = nsym
             name = name.lower()
             sens = sym.sensitivity()
+            
             if sym.type() == "int":
                 if sym.sensitivity() == 1:
-                    counts = self.mechanism.count(srs[name])
-                    counts[counts < 0] = 0
-                    srs[name] = counts
-                    srs = srs.filter(name, ">", self.tau)
+                    if(bounds == "lower"):
+                        lower_bound = srs[name] + self.mechanism.bounds(pct)[0]
+                        srs[name] = lower_bound
+                    elif(bounds == "upper"):
+                        upper_bound = srs[name] + self.mechanism.bounds(pct)[1]
+                        srs[name] = upper_bound
+                    else:
+                        counts = self.mechanism.release(srs[name])
+                        counts[counts < 0] = 0
+                        srs[name] = counts
+                        srs = srs.filter(name, ">", self.tau)
                 elif sens is not None:
-                    srs[name] = self.mechanism.sum_int(srs[name], sens)
+                    mechanism = Laplace(self.epsilon, sens, self.tau)
+                    if(bounds == "lower"):
+                        lower_bound = srs[name] + mechanism.bounds(pct)[0]
+                        srs[name] = lower_bound
+                    elif(bounds == "upper"):
+                        upper_bound = srs[name] + mechanism.bounds(pct)[1]
+                        srs[name] = upper_bound
+                    else:
+                        srs[name] = mechanism.release(srs[name])
             elif sym.type() == "float" and sens is not None:
-                srs[name] = self.mechanism.sum_float(srs[name], sens)
-
+                mechanism = Laplace(self.epsilon, sens, self.tau)
+                if(bounds == "lower"):
+                    lower_bound = srs[name] + mechanism.bounds(pct)[0]
+                    srs[name] = lower_bound
+                elif(bounds == "upper"):
+                    upper_bound = srs[name] + mechanism.bounds(pct)[1]
+                    srs[name] = upper_bound
+                else:
```

Comment:
I'm not sure what this is doing.  It looks like it is returning bounds centered on the exact value for each row of the histogram, which means the midpoint would reveal the exact value.  And bounds should change based on the value of the cell, so bounds should be returned centered on 0, and only once per column.  Also, this could all be done in a single pass.  So, for example:

srs.bounds[name] = mechanism.bounds(pct)
srs[name] = mechanism.release(...

The bounds dict would contain 2-tuples (lower, upper) with the zero-centered bounds that apply to the whole column.

Reply:
Centered bounds on 0 and stored them in srs.bounds to be retrieved in single pass.